src/common/survival_util.h

```diff
@@ -39,47 +92,245 @@ struct AFTParam : public XGBoostParameter<AFTParam> {
+    const double log_y_lower = log(y_lower);
+    const double log_y_upper = log(y_upper);
+    double numerator, denominator, hessian;  // numerator and denominator of hessian
+    CensoringType censor_type;
+    bool z_sign;  // sign of z-score
+
+    if (y_lower == y_upper) {  // uncensored
+      const double z = (log_y_lower - y_pred) / sigma;
+      const double pdf = Distribution::PDF(z);
+      const double grad_pdf = Distribution::GradPDF(z);
+      const double hess_pdf = Distribution::HessPDF(z);
+      censor_type = CensoringType::kUncensored;
+      numerator = -(pdf * hess_pdf - grad_pdf * grad_pdf);
+      denominator = sigma * sigma * pdf * pdf;
+      z_sign = (z > 0);
+    } else {  // censored; now check what type of censorship we have
+      double z_u = 0.0, z_l = 0.0, grad_pdf_u, grad_pdf_l, pdf_u, pdf_l, cdf_u, cdf_l;
+      censor_type = CensoringType::kIntervalCensored;
+      if (isinf(y_upper)) {  // right-censored
+        pdf_u = 0;
+        cdf_u = 1;
+        grad_pdf_u = 0;
+        censor_type = CensoringType::kRightCensored;
+      } else {  // interval-censored or left-censored
+        z_u = (log_y_upper - y_pred) / sigma;
+        pdf_u = Distribution::PDF(z_u);
+        cdf_u = Distribution::CDF(z_u);
+        grad_pdf_u = Distribution::GradPDF(z_u);
+      }
+      if (y_lower <= 0.0) {  // left-censored
+        pdf_l = 0;
+        cdf_l = 0;
+        grad_pdf_l = 0;
+        censor_type = CensoringType::kLeftCensored;
+      } else {  // interval-censored or right-censored
+        z_l = (log_y_lower - y_pred) / sigma;
+        pdf_l = Distribution::PDF(z_l);
+        cdf_l = Distribution::CDF(z_l);
+        grad_pdf_l = Distribution::GradPDF(z_l);
+      }
+      const double cdf_diff = cdf_u - cdf_l;
+      const double pdf_diff = pdf_u - pdf_l;
+      const double grad_diff = grad_pdf_u - grad_pdf_l;
+      const double sqrt_denominator = sigma * cdf_diff;
+      z_sign = (z_u > 0 || z_l > 0);
+      numerator = -(cdf_diff * grad_diff - pdf_diff * pdf_diff);
+      denominator = sqrt_denominator * sqrt_denominator;
+    }
+    hessian = numerator / denominator;
+    if (denominator < aft::kEps && (isnan(hessian) || isinf(hessian))) {
+      hessian = aft::GetLimitHessAtInfPred<Distribution>(censor_type, z_sign, sigma);
+    }
+
+    return aft::Clip(hessian, aft::kMinHessian, aft::kMaxHessian);
+  }
 };
 
+namespace aft {
+
+template <>
+XGBOOST_DEVICE inline double
+GetLimitGradAtInfPred<NormalDistribution>(CensoringType censor_type, bool sign, double sigma) {
+  switch (censor_type) {
+  case CensoringType::kUncensored:
+    return sign ? kMinGradient : kMaxGradient;
+  case CensoringType::kRightCensored:
+    return sign ? kMinGradient : 0.0;
+  case CensoringType::kLeftCensored:
+    return sign ? 0.0 : kMaxGradient;
+  case CensoringType::kIntervalCensored:
+    return sign ? kMinGradient : kMaxGradient;
+  }
+  return nan("");
```

Comment:
No, because we are defining a `__host__ __device__` function, and we cannot use `quiet_NaN` on the device. On the other hand, `nan("")` is valid both on the host and the device.

Reply:
Really?  I just tried it on device code.  What platform are you using?

Reply:
Hmm, it's actually working both on Ubuntu 18.04 LTS and in the CentOS 6 Docker container (GCC 5, CUDA 10.0). There was no indication in the CUDA doc that `std::` constants would be available on the device. I'll update the code to use `std::numeric_limit<float>::quiet_NaN()`.